java/ql/src/experimental/Security/CWE/CWE-749/UnsafeAndroidAccess.ql

```diff
@@ -0,0 +1,135 @@
+  )
+}
+
+/**
+ * Fetch URL method call on the `android.webkit.WebView` object
+ */
+class FetchResourceMethodAccess extends MethodAccess {
+  FetchResourceMethodAccess() {
+    this.getMethod().getDeclaringType() instanceof TypeWebView and
+    (
+      this.getMethod().hasName("loadUrl") or
+      this.getMethod().hasName("postUrl")
+    )
+  }
+}
+
+/**
+ * Method access to external inputs of `android.content.Intent` object
+ */
+class IntentGetExtraMethodAccess extends MethodAccess {
+  IntentGetExtraMethodAccess() {
+    this.getMethod().getName().regexpMatch("get\\w+Extra") and
+    this.getMethod().getDeclaringType() instanceof TypeIntent
+    or
+    this.getMethod().getName().regexpMatch("get\\w+") and
+    this.getQualifier().(MethodAccess).getMethod().hasName("getExtras") and
+    this.getQualifier().(MethodAccess).getMethod().getDeclaringType() instanceof TypeIntent
+  }
+}
+
+/**
+ * Source of fetching urls
+ */
+class UntrustedResourceSource extends RemoteFlowSource {
+  UntrustedResourceSource() {
+    exists(IntentGetExtraMethodAccess ma |
+      this.asExpr().(VarAccess).getVariable().getAnAssignedValue() = ma
+    )
+  }
+
+  override string getSourceType() {
+    result = "user input vulnerable to XSS and sensitive resource disclosure attacks" and
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof CrossOriginAccessMethod and //High precision match of unsafe resource fetching
+      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+    )
+    or
+    result = "user input potentially vulnerable to XSS and sensitive resource disclosure attacks" and
+    not exists(MethodAccess ma |
+      ma.getMethod() instanceof CrossOriginAccessMethod and //High precision match of unsafe resource fetching
+      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+    )
+  }
+}
+
+/**
+ * Holds if `ma` loads url `sink`
+ */
+predicate fetchResource(FetchResourceMethodAccess ma, Expr sink) { sink = ma.getArgument(0) }
+
+/**
+ * Sink of fetching urls
+ */
+class UrlResourceSink extends DataFlow::ExprNode {
+  UrlResourceSink() { fetchResource(_, this.getExpr()) }
+
+  FetchResourceMethodAccess getMethodAccess() { fetchResource(result, this.getExpr()) }
+}
+
+class FetchUntrustedResourceConfiguration extends TaintTracking::Configuration {
+  FetchUntrustedResourceConfiguration() { this = "FetchUntrustedResourceConfiguration" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof UntrustedResourceSource }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof UrlResourceSink }
+}
+
+from DataFlow::PathNode source, DataFlow::PathNode sink, FetchUntrustedResourceConfiguration conf
+where
+  exists(VarAccess webviewVa, MethodAccess getSettingsMa, Variable v |
+    conf.hasFlowPath(source, sink) and
+    sink.getNode().(UrlResourceSink).getMethodAccess().getQualifier() = webviewVa and
+    webviewVa.getVariable().getAnAccess() = getSettingsMa.getQualifier() and
```

Comment:
Should probably constraint `getSettingsMa` to have the right method name. I appreciate `isJSEnabled` ends up constraining it pretty tightly, but this loose arrangement where a thing named `getSettingsMa` is really any method invites a future mistake when editing the code.

Reply:
I've specified the right method name <code>WebViewGetSettingsMethod</code> in the query. Please let me know if this is not what you want.